test/tools/Modules/HelpersDebugger/HelpersDebugger.psm1

```diff
@@ -0,0 +1,217 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Ensure that terminating errors terminate when importing the module.
+trap {throw $_}
+
+# Strict mode FTW.
+Set-StrictMode -Version Latest
+
+# Enable explicit export so that there are no surprises with commands exported from the module.
+Export-ModuleMember
+
+# Grab the internal ScriptPosition property once and re-use it in the ps1xml file
+$internalExtentProperty = [System.Management.Automation.InvocationInfo].GetProperty('ScriptPosition', [System.Reflection.BindingFlags]'NonPublic,Instance')
+
+# A debugger handler that can be used to automatically control the debugger
+$debuggerStopHandler = {
+    param($s, $e)
+    # If we're not handling a debugger stop event during the execution of
+    # Test-Debugger, then simply continue execution
+    if (@(Get-Variable -Name dbgCmdQueue,dbgResults -Scope Script -ErrorAction Ignore).Count -ne 2) {
+        $e.ResumeAction = [System.Management.Automation.DebuggerResumeAction]::Continue
+        return
+    }
+    do {
+        if ($script:dbgCmdQueue.Count -eq 0) {
+            # If there are no more commands to process, continue execution
+            $stringDbgCommand = 'c'
+        } else {
+            $stringDbgCommand = $script:dbgCmdQueue.Dequeue()
+        }
+        $dbgCmd = [System.Management.Automation.PSCommand]::new()
+        $dbgCmd.AddCommand($stringDbgCommand)
+        $output = [System.Management.Automation.PSDataCollection[PSObject]]::new()
+        $result = $Host.Runspace.Debugger.ProcessCommand($dbgCmd, $output)
+        $script:dbgResults += [pscustomobject]@{
+            PSTypeName = 'DebuggerCommandResult'
+            Command = $stringDbgCommand
+            Context = $PSDebugContext
+            Output  = $output
+        }
+    } while ($result -eq $null -or $result.ResumeAction -eq $null)
+    $e.ResumeAction = $result.ResumeAction
+}
+
+# A flag to identify if the debugger handler has been added or not
+$debuggerStopHandlerRegistered = $false
+
+function Register-DebuggerHandler {
+    [CmdletBinding()]
+    param()
+    try {
+        $callerEAP = $ErrorActionPreference
+        # We disable debugger interactivity so that all debugger events go through
+        # the DebuggerStop event only (i.e. breakpoints don't actually generate a
+        # prompt for user interaction)
+        $host.DebuggerEnabled = $false
+        $host.Runspace.Debugger.add_DebuggerStop($script:debuggerStopHandler)
+        $script:debuggerStopHandlerRegistered = $true
+    } catch {
+        Write-Error -ErrorRecord $_ -ErrorAction $callerEAP
+    }
+}
+Export-ModuleMember -Function Register-DebuggerHandler
```

Comment:
Why do we need this is we have FunctionsToExport in manifest?

Reply:
1. Because not all functions are exported.
1. As a best practice, for every module I create, I disable implicit export (by invoking Export-ModuleMember with no arguments at the top of the `.psm1` file, and explicitly exporting the commands that I want exported. This prevents exporting a command accidentally.
1. `FunctionsToExport` is for command discovery, not for command exporting.

Reply:
🤔 if a PSD1 file is provided and used, `Export-ModuleMember` is a no-op.

I don't know why you'd do both, really.

Reply:
`FunctionsToExport` does not export commands. It's for command discovery _only_. Full stop.

When you don't use `Export-ModuleMember`, by default all functions and cmdlets are exported from modules. Whether you want them to or not. Potentially accidentally.

`Export-ModuleMember` is _not_ a no-op. It tells PowerShell that you want to explicitly export commands rather than letting it automatically export commands. This is documented behavior. If you do this at the top of your `.psm1` file, then after you define commands you need to export them, because PowerShell won't. That allows you to control whether commands are exported or not. Internal commands are useful, and the majority of my modules have them. This module has an internal command. It's only internal because of how the module is currently defined.

Reply:
Following a conversation on Slack, it seems PowerShell has changed in recent versions such that `*ToExport` does keep internal functions not listed there internal; however, there are performance gains to be had from explicit management of exports, and I much prefer explicit export as a model in my modules, so I'll leave this in place as is.